### PR TITLE
feat(stream): implement streaming part of now executor

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Summarize your change (**mandatory**)
implement streaming part of now executor, which will return current time `now()`
- How does this PR work? Need a brief introduction for the changed logic (optional)
emit on most barrier
- Describe any limitations of the current code (optional)
frontend part not done

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
#6407

### CODE REVIEW

**Comment by soundOfDestiny** [comment]
we should use `barrier_receiver`

**Comment by soundOfDestiny** [comment]
> we should use `barrier_receiver`

done

File: src/common/src/util/epoch.rs [hunk 1]
```diff
@@ -18,10 +18,12 @@ use std::time::{Duration, SystemTime};
 
 use parse_display::Display;
 
+static UNIX_SINGULARITY_DATE_SEC: u64 = 1_617_235_200;
+
 /// `UNIX_SINGULARITY_DATE_EPOCH` represents the singularity date of the UNIX epoch:
 /// 2021-04-01T00:00:00Z.
 pub static UNIX_SINGULARITY_DATE_EPOCH: LazyLock<SystemTime> =
-    LazyLock::new(|| SystemTime::UNIX_EPOCH + Duration::from_secs(1_617_235_200));
+    LazyLock::new(|| SystemTime::UNIX_EPOCH + Duration::from_secs(UNIX_SINGULARITY_DATE_SEC));
```
Comment (jon-chuang) [inline src/common/src/util/epoch.rs:26]: Do we need LazyLock? Seems both operaands are consts? https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.from_secs

Comment (soundOfDestiny) [reply src/common/src/util/epoch.rs:26]: > Do we need LazyLock? Seems both operaands are consts? https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.from_secs

guess this is because of `SystemTime`

Comment (jon-chuang) [reply src/common/src/util/epoch.rs:26]: Yes, the operation `SystemTime` + `Duration` is unfortunately not const. In particular, const fns in traits such as `Add` still do not exist https://github.com/rust-lang/rfcs/pull/2632

File: dashboard/package.json [hunk 2]
```diff
@@ -27,6 +27,7 @@
     "d3-selection": "^3.0.0",
     "fabric": "^5.2.1",
     "framer-motion": "^6.5.1",
+    "gen-proto": "^0.0.1",
```
Comment (TennyZhuang) [inline dashboard/package.json:30]: Seems unrelated to the PR, please reset the changes

File: dashboard/package-lock.json [hunk 3]
```diff
@@ -21,6 +21,7 @@
         "d3-selection": "^3.0.0",
         "fabric": "^5.2.1",
         "framer-motion": "^6.5.1",
+        "gen-proto": "^0.0.1",
```
Comment (TennyZhuang) [inline dashboard/package-lock.json:24]: ditto

File: src/stream/src/executor/mod.rs [hunk 4]
```diff
@@ -231,6 +233,15 @@ impl Barrier {
         }
     }
 
+    pub fn new_test_barrier_with_prev_epoch(epoch: u64, prev_epoch: u64) -> Self {
```
Comment (TennyZhuang) [inline src/stream/src/executor/mod.rs:236]: How about `with_prev_epoch_for_test`?

Comment (soundOfDestiny) [reply src/stream/src/executor/mod.rs:236]: > How about `with_prev_epoch_for_test`?

this is [wangrunji0408](https://github.com/wangrunji0408)'s suggestion. Anyway I'm to alter it.

Comment (soundOfDestiny) [reply src/stream/src/executor/mod.rs:236]: > How about `with_prev_epoch_for_test`?

`new_test_barrier_with_prev_epoch` is [wangrunji0408](https://github.com/wangrunji0408)'s suggestion. Anyway I'm to alter it.

File: dashboard/package.json (hunk 2 above)
Comment (soundOfDestiny) [reply dashboard/package.json]: > Seems unrelated to the PR, please reset the changes

fixed

File: dashboard/package-lock.json (hunk 3 above)
Comment (soundOfDestiny) [reply dashboard/package-lock.json]: > ditto

fixed

**Comment by TennyZhuang** [comment]
related #6472

File: src/stream/src/executor/now.rs [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+                Some(state_row?)
+            } else {
+                None
+            }
+        };
+
+        let mut last_timestamp = state_row.and_then(|row| row[0].clone());
+
+        while let Some(barrier) = barrier_receiver.recv().await {
+            if !barrier.is_update() && !barrier.is_resume() {
```
Comment (BugenZhao) [inline src/stream/src/executor/now.rs:98]: Suggest maintaining a `paused` state instead of relying on the assumption of "pause -> update -> resume`, in case we have other barriers inside.

File: src/stream/src/executor/now.rs [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        // Consume the first barrier message and initialize state table.
+        let barrier = barrier_receiver
+            .recv()
+            .stack_trace("now_executor_recv_first_barrier")
+            .await
+            .unwrap();
+        state_table.init_epoch(barrier.epoch);
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(barrier);
```
Comment (BugenZhao) [inline src/stream/src/executor/now.rs:83]: So we don't emit the timestamp on the first barrier? Is this expected?

File: src/stream/src/executor/now.rs [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+                    if last_timestamp.is_some() { 2 } else { 1 },
+                );
+                if last_timestamp.is_some() {
+                    let chunk_popped = data_chunk_builder
+                        .append_one_row_from_datums([&last_timestamp].into_iter());
+                    debug_assert!(chunk_popped.is_none());
+                }
+                let data_chunk = data_chunk_builder
+                    .append_one_row_from_datums([&timestamp].into_iter())
+                    .unwrap();
```
Comment (BugenZhao) [inline src/stream/src/executor/now.rs:118]: Directly use `DataChunk::from_rows`?

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM.

File: src/stream/src/executor/now.rs (hunk 6 above)
Comment (soundOfDestiny) [reply src/stream/src/executor/now.rs:83]: > So we don't emit the timestamp on the first barrier? Is this expected?

We can't because no message is permitted before the first barrier.

Comment (soundOfDestiny) [reply src/stream/src/executor/now.rs:83]: > How about emitting the timestamp right after the barrier?

The timestamp will be updated on the second barrier so I think that it is unnecessary.

File: src/stream/src/executor/now.rs (hunk 5 above)
Comment (soundOfDestiny) [reply src/stream/src/executor/now.rs]: > Suggest maintaining a `paused` state instead of relying on the assumption of "pause -> update -> resume`, in case we have other barriers inside.

fixed

File: src/stream/src/executor/now.rs (hunk 7 above)
Comment (soundOfDestiny) [reply src/stream/src/executor/now.rs]: > Directly use `DataChunk::from_rows`?

fixed

**Comment by TennyZhuang** [comment]
Generally LGTM.

I'd prefer to hold the PR for a while, and wait for the frontend part, so that we can test e2e with dynamic filter.

**Comment by soundOfDestiny** [comment]
> Generally LGTM.
> 
> I'd prefer to hold the PR for a while, and wait for the frontend part, so that we can test e2e with dynamic filter.

There will be too much things in this PR:
- backend part
- frontend part
- barrier part because we need to modify meta to let it send barriers to actors which have `Now` Executor

The latter two cannot be placed in a separate PR because they need to use `Now` Executor.

If it is recommended, I will put these into this PR.

**Comment by TennyZhuang** [comment]
> > Generally LGTM.
> > I'd prefer to hold the PR for a while, and wait for the frontend part, so that we can test e2e with dynamic filter.
> 
> There will be too much things in this PR:
> 
> * backend part
> * frontend part
> * barrier part because we need to modify meta to let it send barriers to actors which have `Now` Executor
> 
> The latter two cannot be placed in a separate PR before this because they need to use `Now` Executor.
> 
> If it is recommended, I will put these into this PR.

In multiple PRs, but merge at the same time?

**Comment by soundOfDestiny** [comment]
> > > Generally LGTM.
> > > I'd prefer to hold the PR for a while, and wait for the frontend part, so that we can test e2e with dynamic filter.
> > 
> > 
> > There will be too much things in this PR:
> > 
> > * backend part
> > * frontend part
> > * barrier part because we need to modify meta to let it send barriers to actors which have `Now` Executor
> > 
> > The latter two cannot be placed in a separate PR before this because they need to use `Now` Executor.
> > If it is recommended, I will put these into this PR.
> 
> In multiple PRs, but merge at the same time?

Do you mean that create multiple PRs and other PRs also include codes of `NowExecutor` in this PR?

**Comment by soundOfDestiny** [comment]
> > > Generally LGTM.
> > > I'd prefer to hold the PR for a while, and wait for the frontend part, so that we can test e2e with dynamic filter.
> > 
> > 
> > There will be too much things in this PR:
> > 
> > * backend part
> > * frontend part
> > * barrier part because we need to modify meta to let it send barriers to actors which have `Now` Executor
> > 
> > The latter two cannot be placed in a separate PR before this because they need to use `Now` Executor.
> > If it is recommended, I will put these into this PR.
> 
> In multiple PRs, but merge at the same time?

Therefore let's work on this base PR: #6677  of following multiple PRs first. It defines interfaces of `Now`.

**Comment by jon-chuang** [comment]
> I'd prefer to hold the PR for a while, and wait for the frontend part, so that we can test e2e with dynamic filter.

Actually, if we expect this PR is > 90% correct, then I don't see the harm of merging first, and then once we have e2e with dynamic filter, we can fix any bugs we encounter. And incremental change is easier to review.

IMO Now executor is also independent of dynamic filter.

wdyt @TennyZhuang ?

**Review by jon-chuang — APPROVED** [review]
LGTM

**Review by TennyZhuang — APPROVED** [review]
